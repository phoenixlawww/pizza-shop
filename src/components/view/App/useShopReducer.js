import { useReducer } from 'react'
import { findIndex, omit } from 'lodash'

const defaultState = {
  cartItems: []
}

function reducer (state, action) {
  const { type, ...payload } = action

  switch (action.type) {
    case 'ADD_PRODUCT_TO_CART': {
      const cartItems = [...state.cartItems]
      const product = payload.product

      const currentRecordIndex = findIndex(cartItems, {
        ...omit(product, ['quantity'])
      })

      if (currentRecordIndex !== -1) {
        const currentRecord = cartItems[currentRecordIndex]

        cartItems[currentRecordIndex] = {
          ...currentRecord,
          quantity: currentRecord.quantity + product.quantity
        }
      } else {
        cartItems.push(product)
      }

      return {
        ...state,
        cartItems
      }
    }
    case 'REMOVE_PRODUCT_FROM_CART': {
      return {
        ...state,
        cartItems: state.cartItems.filter((item) => item !== payload.item)
      }
    }
    case 'EMPTY_CART': {
      return {
        ...state,
        cartItems: []
      }
    }
    default: {
      return state
    }
  }
}

export default function useShopReducer (initialState) {
  const [state, dispatch] = useReducer(reducer, {
    ...defaultState,
    ...initialState
  })

  return { state, dispatch }
}
