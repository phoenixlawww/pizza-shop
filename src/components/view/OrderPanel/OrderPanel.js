import { useContext, useEffect, useRef, useState } from 'react'

import pizzaList from 'mock-data/products/pizza.json'
import pizzaOptions from 'mock-data/products/pizza/options.json'

import ProductCard from 'components/shared/ProductCard'
import Cart from 'components/shared/Cart'
import './OrderPanel.scss'
import ShopContext from 'contexts/ShopContext'
import Button, { BUTTON_VARIENTS } from 'components/shared/Button'
import ProductModal from 'components/shared/ProductModal'

function OrderPanel (props) {
  const {
    state: { cartItems },
    dispatch
  } = useContext(ShopContext)

  const [isLoading, setIsLoading] = useState(true)
  const [productList, setProductList] = useState()
  const [optionList, setOptionList] = useState()

  const modalRef = useRef()
  const [modalProduct, setModalProduct] = useState()
  const [modalProductOption, setModalProductOption] = useState()

  useEffect(() => {
    // mock api loading behaviour
    const timeout = setTimeout(() => {
      setProductList(pizzaList.products)
      setOptionList(pizzaOptions.options)
      setIsLoading(false)
    }, 700)

    return () => {
      clearTimeout(timeout)
    }
  }, [])

  return (
    <div className="order-panel">
      {isLoading ? (
        <div className="loading-screen">Loading</div>
      ) : (
        <div className="product-list">
          {productList.map((product, index) => {
            const { productCode, name, image, unitPrice } = product
            return (
              <ProductCard
                className="product-card"
                image={image}
                name={name}
                unitPrice={unitPrice}
                actionBtnLabel="Choose"
                onActionBtnClick={(e) => {
                  setModalProduct(product)
                  setModalProductOption(optionList)
                  modalRef.current.show()
                }}
                key={productCode}
              />
            )
          })}
          <ProductModal
            product={modalProduct}
            productOptions={modalProductOption}
            ref={modalRef}
          />
        </div>
      )}

      <div className="cart-wrapper">
        <Cart className="cart" />
        <Button
          className="empty-cart-btn"
          label="Empty Basket"
          varient={BUTTON_VARIENTS.LINK}
          onClick={(e) => {
            dispatch({
              type: 'EMPTY_CART'
            })
          }}
          disabled={cartItems.length === 0}
        />
      </div>
    </div>
  )
}

export default OrderPanel
