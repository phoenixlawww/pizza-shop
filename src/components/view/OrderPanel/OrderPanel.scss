@use "sass:math";

.order-panel {
  position: relative;
  display: flex;
  flex-direction: row;
  align-items: flex-start;
  justify-content: space-between;
  padding: 20px;
  column-gap: 10px;
  row-gap: 10px;

  > .loading-screen {
    flex: 3;
    align-self: center;
    text-align: center;
  }

  > .product-list {
    display: flex;
    flex: 3;
    flex-direction: row;
    flex-wrap: wrap;
    align-items: stretch;
    justify-content: flex-start;
    column-gap: 10px;
    row-gap: 10px;

    > .product-card {
      flex: 1 0 calc((100% - 10px * 2) / 4);
    }
  }

  > .cart-wrapper {
    display: flex;
    flex: 1;
    flex-direction: column;
    align-items: stretch;
    justify-content: flex-start;

    > .cart {
    }

    > .empty-cart-btn {
      margin: 20px 10px;
    }
  }
}
