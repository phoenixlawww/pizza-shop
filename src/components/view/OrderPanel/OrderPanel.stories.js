import OrderPanel from './OrderPanel'

import ShopContext from 'contexts/ShopContext'
import useShopReducer from 'components/view/App/useShopReducer'

export default {
  component: OrderPanel,
  title: 'View/OrderPanel',
  parameters: {
    layout: 'fullscreen'
  },
  args: {},
  decorators: [
    (Story) => {
      const { state, dispatch } = useShopReducer()

      return (
        <ShopContext.Provider value={{ state, dispatch }}>
          <Story />
        </ShopContext.Provider>
      )
    }
  ]
}

const Template = (args) => <OrderPanel {...args} />

export const Normal = Template.bind({})
Normal.args = {}
