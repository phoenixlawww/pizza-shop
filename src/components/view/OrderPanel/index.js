import OrderPanel from './OrderPanel'
export default OrderPanel
