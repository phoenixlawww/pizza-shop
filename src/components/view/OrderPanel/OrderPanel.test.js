/* eslint-disable react/prop-types */
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { act } from 'react-dom/test-utils'

import ShopContext from 'contexts/ShopContext'
import useShopReducer from 'components/view/App/useShopReducer'

import OrderPanel from './OrderPanel'

import pizzaList from 'mock-data/products/pizza.json'
import pizzaOptions from 'mock-data/products/pizza/options.json'

const actionBtnLabel = 'Choose'
const onActionBtnClick = jest.fn()
const testProps = {
  ...pizzaList.products[0],
  actionBtnLabel,
  onActionBtnClick
}

function OrderPanelWithContext (props) {
  const { defaultValue = {} } = props
  const { state, dispatch } = useShopReducer(defaultValue)

  return (
    <ShopContext.Provider value={{ state, dispatch }}>
      <OrderPanel {...testProps} />
    </ShopContext.Provider>
  )
}

test('renders correctly', () => {
  const { baseElement } = render(<OrderPanelWithContext />)

  expect(baseElement).toMatchSnapshot()
})

test('renders list correctly', () => {
  jest.useFakeTimers()

  const { baseElement } = render(<OrderPanelWithContext />)

  act(() => {
    jest.runAllTimers()
  })

  expect(baseElement).toMatchSnapshot()
})

test('add product into basket successfully', async () => {
  jest.useFakeTimers()

  const { baseElement } = render(<OrderPanelWithContext />)

  act(() => {
    jest.runAllTimers()
  })

  expect(screen.getByText('No Items in your basket')).toBeInTheDocument()

  const productCard = baseElement.querySelector('.product-card .action-btn')
  userEvent.click(productCard)

  const productModal = baseElement.querySelector('.s-product-modal')
  expect(productModal).toBeInTheDocument()

  const topping = productModal.querySelector('input[value="topping-1"]')
  expect(topping).toBeInTheDocument()

  const size = productModal.querySelector('input[value="size-s"]')
  expect(size).toBeInTheDocument()

  const AddToBasketBtn = screen.getByText('Add To Basket')
  expect(AddToBasketBtn).toBeDisabled()

  await act(async () => {
    userEvent.click(topping)
  })

  expect(AddToBasketBtn).toBeDisabled()

  await act(async () => {
    userEvent.click(size)
  })

  expect(AddToBasketBtn).not.toBeDisabled()

  await act(async () => {
    userEvent.click(AddToBasketBtn)
  })

  expect(productModal).not.toBeInTheDocument()

  expect(screen.getByText('1 x Pizza A')).toBeInTheDocument()
})

test('remove product from basket successfully', async () => {
  jest.useFakeTimers()

  const { baseElement } = render(
    <OrderPanelWithContext
      defaultValue={{
        cartItems: [
          {
            ...pizzaList.products[0],
            options: [
              pizzaOptions.options[0].subOptions[0],
              pizzaOptions.options[1].subOptions[0]
            ],
            quantity: 1
          }
        ]
      }}
    />
  )

  act(() => {
    jest.runAllTimers()
  })

  expect(screen.getByText('1 x Pizza A')).toBeInTheDocument()

  const removeProductBtn = baseElement.querySelector('.item-delete-btn')

  await act(async () => {
    userEvent.click(removeProductBtn)
  })

  expect(screen.getByText('No Items in your basket')).toBeInTheDocument()
})

test('empty basket successfully', async () => {
  jest.useFakeTimers()

  render(
    <OrderPanelWithContext
      defaultValue={{
        cartItems: [
          {
            ...pizzaList.products[0],
            options: [
              pizzaOptions.options[0].subOptions[0],
              pizzaOptions.options[1].subOptions[0]
            ],
            quantity: 1
          }
        ]
      }}
    />
  )

  act(() => {
    jest.runAllTimers()
  })

  expect(screen.getByText('1 x Pizza A')).toBeInTheDocument()

  const emptyBasketBtn = screen.getByText('Empty Basket')

  await act(async () => {
    userEvent.click(emptyBasketBtn)
  })

  expect(screen.getByText('No Items in your basket')).toBeInTheDocument()
})
