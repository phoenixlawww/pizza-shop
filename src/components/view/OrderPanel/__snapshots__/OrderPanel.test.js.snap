// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<body>
  <div>
    <div
      class="order-panel"
    >
      <div
        class="loading-screen"
      >
        Loading
      </div>
      <div
        class="cart-wrapper"
      >
        <div
          class="s-cart cart"
        >
          <div
            class="cart-empty"
          >
            No Items in your basket
          </div>
          <div
            class="cart-total"
          >
            <div>
              Total
            </div>
            <div
              class="cart-amount"
            >
              $0
            </div>
          </div>
          <button
            class="s-button s-button-primary"
            disabled=""
          >
            Checkout
          </button>
        </div>
        <button
          class="s-button empty-cart-btn s-button-link"
          disabled=""
        >
          Empty Basket
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`renders list correctly 1`] = `
<body>
  <div>
    <div
      class="order-panel"
    >
      <div
        class="product-list"
      >
        <div
          class="s-product-card product-card"
        >
          <div
            class="product-img-wrapper"
          >
            <img
              alt=""
              src="/pizza.jpg"
            />
          </div>
          <h4
            class="product-name"
          >
            Pizza A - $99
          </h4>
          <button
            class="s-button action-btn s-button-primary"
          >
            Choose
          </button>
        </div>
        <div
          class="s-product-card product-card"
        >
          <div
            class="product-img-wrapper"
          >
            <img
              alt=""
              src="/pizza.jpg"
            />
          </div>
          <h4
            class="product-name"
          >
            Pizza B - $99
          </h4>
          <button
            class="s-button action-btn s-button-primary"
          >
            Choose
          </button>
        </div>
        <div
          class="s-product-card product-card"
        >
          <div
            class="product-img-wrapper"
          >
            <img
              alt=""
              src="/pizza.jpg"
            />
          </div>
          <h4
            class="product-name"
          >
            Pizza C - $99
          </h4>
          <button
            class="s-button action-btn s-button-primary"
          >
            Choose
          </button>
        </div>
        <div
          class="s-product-card product-card"
        >
          <div
            class="product-img-wrapper"
          >
            <img
              alt=""
              src="/pizza.jpg"
            />
          </div>
          <h4
            class="product-name"
          >
            Pizza D - $99
          </h4>
          <button
            class="s-button action-btn s-button-primary"
          >
            Choose
          </button>
        </div>
        <div
          class="s-product-card product-card"
        >
          <div
            class="product-img-wrapper"
          >
            <img
              alt=""
              src="/pizza.jpg"
            />
          </div>
          <h4
            class="product-name"
          >
            Pizza E - $99
          </h4>
          <button
            class="s-button action-btn s-button-primary"
          >
            Choose
          </button>
        </div>
        <div
          class="s-product-card product-card"
        >
          <div
            class="product-img-wrapper"
          >
            <img
              alt=""
              src="/pizza.jpg"
            />
          </div>
          <h4
            class="product-name"
          >
            Pizza F - $99
          </h4>
          <button
            class="s-button action-btn s-button-primary"
          >
            Choose
          </button>
        </div>
      </div>
      <div
        class="cart-wrapper"
      >
        <div
          class="s-cart cart"
        >
          <div
            class="cart-empty"
          >
            No Items in your basket
          </div>
          <div
            class="cart-total"
          >
            <div>
              Total
            </div>
            <div
              class="cart-amount"
            >
              $0
            </div>
          </div>
          <button
            class="s-button s-button-primary"
            disabled=""
          >
            Checkout
          </button>
        </div>
        <button
          class="s-button empty-cart-btn s-button-link"
          disabled=""
        >
          Empty Basket
        </button>
      </div>
    </div>
  </div>
</body>
`;
